funcbench/main.go

```diff
@@ -0,0 +1,362 @@
+			if postCommentErr := ghClient.postComment(comment); postCommentErr != nil {
+				log.Fatalf("An error: %v occured while processing error: %v", postCommentErr, err)
+			}
+			log.Fatalln(err)
+		}
+
+		if strings.Count(unformatedComparisonTable, "\n") < 2 { // Maybe delete as this error gets caught sooner
+			comment := fmt.Sprintf("Error: `go test` did not match any `BenchmarkXxx` functions. %s", ghClient.logLink)
+			if postCommentErr := ghClient.postComment(comment); postCommentErr != nil {
+				log.Fatalf("An error: %v occured while processing error: %v", postCommentErr, err)
+			}
+		}
+		if err := ghClient.postComment(formatFinalComment(unformatedComparisonTable)); err != nil {
+			log.Fatalln(err)
+		}
+
+	default:
+		log.Fatalln("only issue_comment event is supported")
+	}
+}
+
+func newGitHubClient(event *github.IssueCommentEvent) *gitHubClient {
+	owner := *event.GetRepo().Owner.Login
+	repo := *event.GetRepo().Name
+	latestCommitHash := os.Getenv("GITHUB_SHA")
+
+	return &gitHubClient{
+		client:           newClient(os.Getenv("GITHUB_TOKEN")),
+		owner:            owner,
+		repo:             repo,
+		prNumber:         *event.GetIssue().Number,
+		latestCommitHash: latestCommitHash,
+		logLink:          fmt.Sprintf("Full logs at: https://github.com/%s/%s/commit/%s/checks", owner, repo, latestCommitHash),
+	}
+}
+
+func (c *gitHubClient) postComment(comment string) error {
+	issueComment := &github.IssueComment{Body: github.String(comment)}
+	_, _, err := c.client.Issues.CreateComment(context.Background(), c.owner, c.repo, c.prNumber, issueComment)
+	return err
+}
+
+func newGitClient(event *github.IssueCommentEvent) (*gitClient, error) {
+	branch, err := readFile("/github/home/BRANCH")
+	c := &gitClient{
+		branch:   branch,
+		owner:    *event.GetRepo().Owner.Login,
+		repo:     *event.GetRepo().Name,
+		prNumber: *event.GetIssue().Number,
+	}
+
+	log.Println("BRANCH / branch = ", c.branch,
+		"PR Number = ", c.prNumber,
+		"Owner ", c.owner,
+		"Repo ", c.repo)
+	return c, err
+}
+
+func (c *gitClient) checkoutPullRequestFile(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	if info.IsDir() || strings.Contains(path, ".git/") || strings.HasSuffix(info.Name(), "test.go") {
+		return nil
+	}
+
+	data, err := exec.Command("git", "checkout", "--", path).CombinedOutput()
+	if err != nil {
+		log.Println(string(data))
+
+		// New file appeared and is not tracked, then it can be removed.
+		if strings.Contains(string(data), "did not match any file") {
+			data, err = exec.Command("rm", path).CombinedOutput()
+			if err != nil {
+				return errors.Errorf("Error: %v; Command out: %s", err, string(data))
+			}
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+func (c *gitClient) cloneRepository() error {
+	_, err := execCommand("git", "clone", fmt.Sprintf("https://github.com/%s/%s.git", c.owner, c.repo))
+	return err
+}
+
+func (c *gitClient) checkoutToPullRequest() error {
```

Comment:
can you put a comment for this function what it does exactly as I don't quite understand it and it will be a good reminder for when we read it in few weeks time.